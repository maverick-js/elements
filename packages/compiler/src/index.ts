export * from './plugins';
export * from './transformer';
