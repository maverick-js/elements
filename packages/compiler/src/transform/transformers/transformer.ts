import type ts from 'typescript';

import type { AstNode } from '../../parse/ast';
import type { TransformContext } from '../transform';

export interface Transform {
  (data: TransformData): ts.SourceFile;
}

export interface TransformData {
  sourceFile: ts.SourceFile;
  nodes: AstNode[];
  ctx: TransformContext;
}

export interface StateTransform<State> {
  (node: AstNode, state: State): StateTransformResult;
}

export type StateTransformResult = ts.Expression | Array<ts.Expression | ts.Statement> | undefined;
