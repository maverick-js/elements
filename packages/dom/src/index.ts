export * from './insert';
export * from './reconcile';
export * from './render';
export * from './runtime';
export * from './walker';

export * from './components/for';
export * from './components/fragment';
export * from './components/host';
export * from './components/portal';
