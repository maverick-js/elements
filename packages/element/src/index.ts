export * from './create';
export * from './element';
export * from './types';
export * from './define';
export * from './symbols';
