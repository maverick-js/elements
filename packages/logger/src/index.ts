export * from './logger';
export * from './ms';
