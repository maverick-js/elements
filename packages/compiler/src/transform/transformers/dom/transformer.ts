import {
  $,
  createImports,
  isJsxRootNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { Scope } from '../../../parse/ast';
import { parse } from '../../../parse/parse';
import { setupCustomElements } from '../shared/element';
import { Variables } from '../shared/variables';
import type { Transform, TransformData } from '../transformer';
import type { DomRuntime } from './runtime';
import { DomTransformState } from './state';
import { transform } from './transform';

export interface DomTransformOptions {
  /**
   * Whether to mark components with a static `element` property declaration to be
   * built as a custom element.
   */
  customElements?: boolean;
  /**
   * When set to `true`, the DOM runtime will connect the server-rendered HTML with the JS on
   * the client, making it interactive.
   *
   * The runtime does not require the component to be rendered on the server first but it's
   * recommended for the best performance.
   */
  hydratable?: boolean;
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createDomTransform(options?: DomTransformOptions): Transform {
  return (data) => domTransform(data, options);
}

export function domTransform(
  { sourceFile }: TransformData,
  { customElements, hydratable, delegateEvents }: DomTransformOptions = {},
) {
  let state = new DomTransformState(null, { hydratable }),
    seenImports = false;

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast, new Scope())) : $.null;

      resetUniqueIdCount();

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.runtime));
  }

  const statements: ts.Statement[] = [];

  const templates = createTemplateVariables(state);

  if (state.runtime.identifiers.length > 0) {
    statements.push(state.runtime.toImportDeclaration());
  }

  if (templates) statements.push(templates.toStatement());

  if (state.vars.module.length > 0) {
    statements.push(state.vars.module.toStatement());
  }

  const updatedSourceFile = $.updateSourceFile(transformedSourceFile, [
    ...imports,
    ...statements,
    ...state.renders,
    ...body,
  ]);

  return customElements
    ? setupCustomElements(updatedSourceFile, registerCustomElement, [createElementImport()])
    : updatedSourceFile;
}

const registerCustomElementExpression = $.call($.id('$$_define_custom_element'), [$.id('this')]);

function registerCustomElement() {
  return registerCustomElementExpression;
}

function createElementImport() {
  return createImports([$.id('$$_define_custom_element')], '@maverick-js/element');
}

function createTemplateVariables(state: DomTransformState) {
  const templates: [template: ts.Identifier, html: string, importNodes: boolean][] = [],
    vars = new Variables();

  state.walk(({ template, html, importNodes }) => {
    if (template && html) templates.push([template, html, importNodes]);
  });

  for (const [template, html, importNodes] of templates) {
    vars.create(template, state.runtime.createTemplate($.string(html), importNodes));
  }

  // Dedupe templates.
  for (let i = 0; i < templates.length; i++) {
    const templateA = templates[i];
    for (let j = i + 1; j < templates.length; j++) {
      const templateB = templates[j];
      if (templateA[1] === templateB[1] && templateA[2] === templateB[2]) {
        vars.update(templateB[0], templateA[0]);
      }
    }
  }

  return templates.length > 0 ? vars : null;
}

export function createDelegateEventsStatement(events: Set<string>, runtime: DomRuntime) {
  return $.createExpressionStatement(
    runtime.delegateEvents($.array(Array.from(events).map((type) => $.string(type)))),
  );
}
