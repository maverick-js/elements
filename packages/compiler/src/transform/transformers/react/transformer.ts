import { isArray } from '@maverick-js/std';
import {
  $,
  addStatementsToBlock,
  type BlockUpdate,
  isJsxRootNode,
  replaceTsNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
  updateBlock,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { parse } from '../../../parse/parse';
import { createDelegateEventsStatement } from '../dom/transformer';
import { returnLastExpression } from '../shared/factory';
import type { Transform, TransformData } from '../transformer';
import { ReactTransformState } from './state';
import { transform } from './transform';

export interface ReactTransformOptions {
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createReactTransform(options?: ReactTransformOptions): Transform {
  return (data) => reactTransform(data, options);
}

export function reactTransform(
  { sourceFile }: TransformData,
  { delegateEvents }: ReactTransformOptions = {},
) {
  let state = new ReactTransformState(null),
    seenImports = false,
    blocks: BlockUpdate[][] = [];

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (ts.isBlock(node) || ts.isModuleBlock(node)) {
      const updates: BlockUpdate[] = [];

      blocks.push(updates);
      const block = ts.visitEachChild(node, visit, this);
      blocks.pop();

      return updateBlock(block, updates);
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast)) : $.null,
        currentBlock = blocks.at(-1);

      resetUniqueIdCount();

      if (currentBlock) {
        if (isArray(result)) {
          const statements = result.slice(0, -1);

          currentBlock.push({
            type: 'add',
            statements,
            before: node,
          });

          currentBlock.push({
            type: 'replace',
            old: node,
            new: result.at(-1)!,
          });
        } else {
          currentBlock.push({
            type: 'replace',
            old: node,
            new: result,
          });
        }

        // Return original block and transform above.
        return node;
      }

      if (isArray(result)) {
        returnLastExpression(result);
        return $.selfInvokedFn(result);
      }

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.domRuntime));
  }

  const statements: ts.Statement[] = [];

  const runtimes = [state.runtime, state.domRuntime, state.ssrRuntime];
  for (const runtime of runtimes) {
    if (runtime.identifiers.length > 0) {
      statements.push(runtime.toImportDeclaration());
    }
  }

  if (state.module.vars.length > 0) {
    statements.push(state.module.vars.toStatement());
  }

  return $.updateSourceFile(sourceFile, [
    ...imports,
    ...statements,
    ...state.module.block,
    ...body,
  ]);
}
