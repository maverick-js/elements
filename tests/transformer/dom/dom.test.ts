import { transform } from 'src/transformer';

const t = (code: string) => transform(code).code;

it('should return sourcemap', () => {
  const result = transform(`<div></div>`, { filename: 'foo.tsx', sourcemap: true });
  expect(result.map).toBeDefined();
  expect(result.map!.sources[0]).toBe('foo.tsx');
});

it('should compile single JSX node', () => {
  const result = t(`<div></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile single self-closing JSX node', () => {
  const result = t(`<div />`);
  expect(result).toMatchSnapshot();
});

it('should compile multiple JSX nodes', () => {
  const result = t(`<div><span id="a"></span><span id="b"></span></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile fragment', () => {
  const result = t(`<><div id="a"></div><div id="b"></div></>`);
  expect(result).toMatchSnapshot();
});

it('should compile nested fragment', () => {
  const result = t(`<div id="root"><><div id="a"></div><div id="b"></div></></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile shorthand boolean attribute', () => {
  const result = t(`<button disabled />`);
  expect(result).toMatchSnapshot();
});

it('should compile static attributes', () => {
  const result = t(`<div class="foo bar" style="baz daz"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static attribute (number)', () => {
  const result = t(`<div foo={10}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static attribute (boolean)', () => {
  const result = t(`<div foo={true} bar={false}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile static attribute (template string)', () => {
  const result = t(`<div foo={\`bar-baz\`} ></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile spread', () => {
  const result = t(`<div {...props} ></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile SVG spread', () => {
  const result = t(`<svg {...props}></svg>`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic attribute', () => {
  const result = t(`<div foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable attribute', () => {
  const result = t(`<div foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile innerHTML expression', () => {
  const result = t(`<div $prop:innerHTML="baz"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $prop expression', () => {
  const result = t(`<div $prop:fooBar="baz"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile shorthand $prop boolean', () => {
  const result = t(`<button $prop:disabled />`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic $prop expression', () => {
  const result = t(`<div $prop:foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable $prop expression', () => {
  const result = t(`<div $prop:foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $class expression', () => {
  const result = t(`<div $class:foo={true}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable $class expression', () => {
  const result = t(`<div $class:foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $style expression', () => {
  const result = t(`<div $style:foo="bar"></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile dynamic $style expression', () => {
  const result = t(`<div $style:foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should group multiple static $style expressions', () => {
  const result = t(
    `<div style="foo: a; " $style:foo={"b"} $style:bar={true} $style:baz={id}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should compile observable $style expression', () => {
  const result = t(`<div $style:foo={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $cssvar expression', () => {
  const result = t(`<div $cssvar:foo={10}></div>`);
  expect(result).toMatchSnapshot();
});

it('should group multiple static $cssvar expressions', () => {
  const result = t(
    `<div style="pre: 10" $cssvar:foo={10} $cssvar:bar={'align-content'} $cssvar:baz={id}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should group multiple static $style and $cssvar expressions', () => {
  const result = t(
    `<div style="pre: 10" $style:baz={\`content\`} $style:boo={20} $cssvar:foo={10} $cssvar:bar={'align-content'}></div>`,
  );
  expect(result).toMatchSnapshot();
});

it('should compile dynamic $cssvar expression', () => {
  const result = t(`<div $cssvar:foo={id}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable $cssvar expression', () => {
  const result = t(`<div $cssvar:foo-bar={id()}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $use expression', () => {
  const result = t(`<div $use:directive={[1, 2, 3]}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $ref expression', () => {
  const result = t(`<div $ref={(el) => {}}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $ref expression that uses array', () => {
  const result = t(`<div $ref={[(el) => {}, (el) => {}]}></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile $on expression', () => {
  const result = t(`<div $on:foo={(e) => {}} />`);
  expect(result).toMatchSnapshot();
});

it('should compile multiple $on expression', () => {
  const result = t(`<div $on:foo={(e) => {}} $on:foo={(e) =>{}} />`);
  expect(result).toMatchSnapshot();
});

it('should compile $on_capture expression', () => {
  const result = t(`<div $on_capture:foo={(e) => {}} />`);
  expect(result).toMatchSnapshot();
});

it('should compile native $on expression', () => {
  const result = t(`<div $on:click={(e) => {}} />`);
  expect(result).toMatchSnapshot();
});

it('should compile child expression', () => {
  const result = t(`<div>{id}</div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable child expression', () => {
  const result = t(`<div>{id() + 10}</div>`);
  expect(result).toMatchSnapshot();
});

it('should compile conditional element expression ', () => {
  const result = t(`<div id="a">{id > 10 && <div id="b"></div>}</div>`);
  expect(result).toMatchSnapshot();
});

it('should compile observable conditional element expression ', () => {
  const result = t(`<div id="a">{id() > 10 && <div id="b" $on:click={id()}></div>}</div>`);
  expect(result).toMatchSnapshot();
});

it('shoud merge duplicate templates', () => {
  const result = t(`
<div></div>
<div></div>
<div>Foo</div>
<div>Foo</div>
<div>Bar</div>
<span></span>
<span>Foo</span>
<span>Foo</span>
<span>Foo</span>
<span>Bar</span>
`);
  expect(result).toMatchSnapshot();
});
