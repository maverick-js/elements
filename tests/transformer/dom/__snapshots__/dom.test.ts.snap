// Vitest Snapshot v1

exports[`shoud merge duplicate templates 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_templ,
  $$_templ_3 = /* #__PURE__ */ $$_create_template(\`<div>Foo</div>\`),
  $$_templ_4 = /* #__PURE__ */ $$_templ_3,
  $$_templ_5 = /* #__PURE__ */ $$_create_template(\`<div>Bar</div>\`),
  $$_templ_6 = /* #__PURE__ */ $$_create_template(\`<span></span>\`),
  $$_templ_7 = /* #__PURE__ */ $$_create_template(\`<span>Foo</span>\`),
  $$_templ_8 = /* #__PURE__ */ $$_templ_7,
  $$_templ_9 = /* #__PURE__ */ $$_templ_7,
  $$_templ_10 = /* #__PURE__ */ $$_create_template(\`<span>Bar</span>\`);

$$_clone($$_templ, 1 /* ELEMENT */)
$$_clone($$_templ_2, 1 /* ELEMENT */)
$$_clone($$_templ_3, 1 /* ELEMENT */)
$$_clone($$_templ_4, 1 /* ELEMENT */)
$$_clone($$_templ_5, 1 /* ELEMENT */)
$$_clone($$_templ_6, 1 /* ELEMENT */)
$$_clone($$_templ_7, 1 /* ELEMENT */)
$$_clone($$_templ_8, 1 /* ELEMENT */)
$$_clone($$_templ_9, 1 /* ELEMENT */)
$$_clone($$_templ_10, 1 /* ELEMENT */)
"
`;

exports[`should compile $class expression 1`] = `
"import { $$_clone, $$_class, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_class($$_el, \\"foo\\", true);

  return $$_el;
})()"
`;

exports[`should compile $cssvar expression 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div style=\\"--foo: 10\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile $on expression 1`] = `
"import { $$_clone, $$_listen, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_listen($$_el, \\"foo\\", (e) => {});

  return $$_el;
})()"
`;

exports[`should compile $oncapture expression 1`] = `
"import { $$_clone, $$_listen, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_listen($$_el, \\"foo\\", (e) => {}, 1 /* CAPTURE */);

  return $$_el;
})()"
`;

exports[`should compile $prop expression 1`] = `
"import { $$_clone, $$_prop, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_prop($$_el, \\"fooBar\\", \\"baz\\");

  return $$_el;
})()"
`;

exports[`should compile $ref expression 1`] = `
"import { $$_clone, $$_ref, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_ref($$_el, (el) => {});

  return $$_el;
})()"
`;

exports[`should compile $ref expression that uses array 1`] = `
"import { $$_clone, $$_ref, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_ref($$_el, [(el) => {}, (el) => {}]);

  return $$_el;
})()"
`;

exports[`should compile $style expression 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div style=\\"foo: bar\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile $use expression 1`] = `
"import { $$_clone, $$_directive, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_directive($$_el, directive, [1, 2, 3]);

  return $$_el;
})()"
`;

exports[`should compile SVG spread 1`] = `
"import { $$_clone, $$_spread, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<svg></svg>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_spread($$_el, props);

  return $$_el;
})()"
`;

exports[`should compile child expression 1`] = `
"import { $$_clone, $$_insert, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_insert($$_el, id);

  return $$_el;
})()"
`;

exports[`should compile conditional element expression  1`] = `
"import { $$_clone, $$_create_template, $$_insert } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div id=\\"a\\"></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_create_template(\`<div id=\\"b\\"></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_insert($$_el, id > 10 && $$_clone($$_templ_2, 1 /* ELEMENT */));

  return $$_el;
})()"
`;

exports[`should compile dynamic $cssvar expression 1`] = `
"import { $$_clone, $$_cssvar, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_cssvar($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile dynamic $prop expression 1`] = `
"import { $$_clone, $$_prop, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_prop($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile dynamic $style expression 1`] = `
"import { $$_clone, $$_style, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_style($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile dynamic attribute 1`] = `
"import { $$_clone, $$_attr, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_attr($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile fragment 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div id=\\"a\\"></div><div id=\\"b\\"></div>\`);
$$_clone($$_templ)"
`;

exports[`should compile innerHTML expression 1`] = `
"import { $$_clone, $$_inner_html, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_inner_html($$_el, \\"baz\\");

  return $$_el;
})()"
`;

exports[`should compile multiple $on expression 1`] = `
"import { $$_clone, $$_listen, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_listen($$_el, \\"foo\\", (e) => {});
  $$_listen($$_el, \\"foo\\", (e) => {});

  return $$_el;
})()"
`;

exports[`should compile multiple JSX nodes 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><span id=\\"a\\"></span><span id=\\"b\\"></span></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile native $on expression 1`] = `
"import { $$_clone, $$_listen, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_listen($$_el, \\"click\\", (e) => {});

  return $$_el;
})()"
`;

exports[`should compile nested fragment 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div id=\\"root\\"><div id=\\"a\\"></div><div id=\\"b\\"></div></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile observable $class expression 1`] = `
"import { $$_clone, $$_class, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_class($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile observable $cssvar expression 1`] = `
"import { $$_clone, $$_cssvar, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_cssvar($$_el, \\"foo-bar\\", id);

  return $$_el;
})()"
`;

exports[`should compile observable $prop expression 1`] = `
"import { $$_clone, $$_prop, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_prop($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile observable $style expression 1`] = `
"import { $$_clone, $$_style, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_style($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile observable attribute 1`] = `
"import { $$_clone, $$_attr, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_attr($$_el, \\"foo\\", id);

  return $$_el;
})()"
`;

exports[`should compile observable child expression 1`] = `
"import { $$_clone, $$_insert, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_insert($$_el, () => id() + 10);

  return $$_el;
})()"
`;

exports[`should compile observable conditional element expression  1`] = `
"import { $$_clone, $$_listen, $$_create_template, $$_insert } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div id=\\"a\\"></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_create_template(\`<div id=\\"b\\"></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_insert($$_el, () =>
    id() > 10 && (() => {
      const $$_root = $$_clone($$_templ_2),
        $$_el = $$_root.firstChild;

      $$_listen($$_el, \\"click\\", id());

      return $$_el;
    })());

  return $$_el;
})()"
`;

exports[`should compile shorthand $prop boolean 1`] = `
"import { $$_clone, $$_prop, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<button></button>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_prop($$_el, \\"disabled\\", true);

  return $$_el;
})()"
`;

exports[`should compile shorthand boolean attribute 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<button disabled=\\"\\"></button>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile single JSX node 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile single self-closing JSX node 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile spread 1`] = `
"import { $$_clone, $$_spread, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_spread($$_el, props);

  return $$_el;
})()"
`;

exports[`should compile static attribute (boolean) 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div foo=\\"true\\" bar=\\"false\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile static attribute (number) 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div foo=\\"10\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile static attribute (template string) 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div foo=\\"bar-baz\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should compile static attributes 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div class=\\"foo bar\\" style=\\"baz daz\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should group multiple static $cssvar expressions 1`] = `
"import { $$_clone, $$_cssvar, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div style=\\"pre: 10;--foo: 10;--bar: align-content\\"></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_cssvar($$_el, \\"baz\\", id);

  return $$_el;
})()"
`;

exports[`should group multiple static $style and $cssvar expressions 1`] = `
"import { $$_create_template, $$_clone } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div style=\\"pre: 10;baz: content;boo: 20;--foo: 10;--bar: align-content\\"></div>\`);
$$_clone($$_templ, 1 /* ELEMENT */)"
`;

exports[`should group multiple static $style expressions 1`] = `
"import { $$_clone, $$_style, $$_create_template } from \\"maverick.js/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div style=\\"foo: a;foo: b;bar: true\\"></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_el = $$_root.firstChild;

  $$_style($$_el, \\"baz\\", id);

  return $$_el;
})()"
`;
